[MIOpen] Convert cba inference ctest to GTest

kokolchin opened the pull request:
## Motivation

This PR unifies the migration of cba_inference and cbna_inference tests from the legacy CTest framework to Google Test (GTest). This transition improves test consistency, provides superior failure isolation, and removes the dependency on the deprecated test_driver mechanism.

## Technical Details

- Branch Merging: This PR combines the refactored CBA (Convolution-Bias-Activation) and CBNA (Convolution-Bias-BatchNorm-Activation) inference tests into a single feature branch.
- Framework Migration: Legacy tests are converted to a modern GTest architecture utilizing standard fixtures and parameterized test suites.
- Test Case Parity: Test points are generated using get_inputs() and get_weights() in the exact Cartesian product sequence used by the legacy driver.
- Reporting Transparency: The suite utilizes GTEST_SKIP() to accurately report cases where fusion plans are unsupported by the current backend solvers or where dimension requirements are not met, matching the functional logic of the original driver.
- Performance: Batch sizes are restricted to 1 to align with legacy driver defaults, ensuring efficient CPU reference calculations.

## Test Plan

- Verify that the total test count and iteration sequence for both CBA and CBNA match the original CTest implementation.
- Confirm that the number of successful verifications in GTest is identical to the "Successful Test Count" reported by the legacy driver.

## Test Result
- The total test count and iteration sequence for both CBA and CBNA match the original CTest implementation.

dpolominDXC commented on projects/miopen/test/gtest/cba_inference.cpp:
```diff
@@ -0,0 +1,603 @@
+    }
+};
+
+template <typename T>
```
Does it have to be templated? `T` doesn't participate anywhere in the function

kokolchin replied:
No need. Removed.

Vsevolod1983 commented on projects/miopen/test/gtest/cba_inference.cpp:
```diff
@@ -0,0 +1,603 @@
+};
+
+template <typename T>
+std::vector<CbaTestCase> GetCbaTestCases()
```
I can see only one test case here, is it correct ?
Please ensure we haven't missed test cases comparing to c-test

kokolchin replied:
Confirmed parity: this GTest covers the same set of cases as the original CTest (87 cases).

Vsevolod1983 commented on projects/miopen/test/gtest/cba_inference.cpp:
```diff
@@ -0,0 +1,603 @@
+    void SetUp() override
+    {
+        prng::reset_seed();
+        if(!IsTestSupportedByDevice(Gpu::All))
```
What the intention of this check ?

kokolchin replied:
Removed.

kokolchin replied:
Removed.

Vsevolod1983 commented on projects/miopen/test/gtest/cba_inference.cpp:
```diff
@@ -0,0 +1,603 @@
+                                                               ptr_activdesc.get(),
+                                                               workspace_size};
+
+                    auto cpu_result = verifier.cpu();
```
Looks like we have partly duplicated code 3 times ?

kokolchin replied:
Corrected by introducing the VerifyAndValidate<T>() helper function.

kokolchin replied:
Corrected.

randyspauldingamd commented on projects/miopen/test/gtest/cbna_inference.cpp:
```diff
@@ -0,0 +1,357 @@
+        auto wei_dev  = handle.Write(weights.data);
+        auto b_dev    = handle.Write(bias.data);
+        auto out_dev  = handle.Write(rout.data);
+        auto sc_dev   = handle.Write(bnscale.data);
+        auto sh_dev   = handle.Write(bnbias.data);
+        auto m_dev    = handle.Write(estMean.data);
+        auto v_dev    = handle.Write(estVariance.data);
```
Please change these back to the original names (`<member>_dev`) for clarity

kokolchin replied:
Done — variable names were reverted to the original <member>_dev style for clarity.

randyspauldingamd commented on projects/miopen/test/gtest/cbna_inference.cpp:
```diff
@@ -0,0 +1,357 @@
+    filter.dilations[0] = test_case.pads_strides_dilations[4];
+    filter.dilations[1] = test_case.pads_strides_dilations[5];
+
+    if(input_c == wei_c)
```
There's a tacit PASS here that is incorrect. Let's make this a hard check and return as soon as possible since there isn't a reason to continue. Please remove this if block, and see the comments on line 259 and 301.

kokolchin replied:
Addressed. The tacit pass path was removed and control flow was reworked to fail/skip early as requested.

randyspauldingamd commented on projects/miopen/test/gtest/cbna_inference.cpp:
```diff
@@ -0,0 +1,357 @@
+        miopenCreateOpBatchNormInference(ptr_fusionplan.get(), &bnOp, bnmode, &bnscale.desc);
+        miopenCreateOpActivationForward(ptr_fusionplan.get(), &activOp, miopenActivationRELU);
+
+        miopenStatus_t miopenError = miopenCompileFusionPlan(&handle, ptr_fusionplan.get());
```
For the sake of simplicity we don't need to check every operation, but do add a check and skip here. We'll also check for expected failures here:
```
    // Failure to catch an invalid shape results in test failure.
    EXPECT_PRED2([](bool valid, miopenStatus_t status) {
       return valid || (status != miopenStatusSuccess);
    }, valid_test_case, miopenError) << "Test case contains incompatible shapes. The library is expected to reject this; however, it did not.";

     if(miopenError != miopenStatusSuccess)
     {
         GTEST_SKIP() << "No supported fusion solver found for: " << test_case;
     }
```

kokolchin replied:
Updated to enforce expected-failure behavior for invalid shapes and to skip unsupported (but otherwise valid) solver cases.

randyspauldingamd commented on projects/miopen/test/gtest/cbna_inference.cpp:
```diff
@@ -0,0 +1,357 @@
+        miopenStatus_t miopenError = miopenCompileFusionPlan(&handle, ptr_fusionplan.get());
+
+        size_t workspace_size = 0;
+        if(miopenError == miopenStatusSuccess)
```
After the other changes above, this `if` should be redundant. Remove it.

kokolchin replied:
Removed.

randyspauldingamd commented on projects/miopen/test/gtest/cbna_inference.cpp:
```diff
@@ -0,0 +1,357 @@
+    filter.strides[1]   = test_case.pads_strides_dilations[3];
+    filter.dilations[0] = test_case.pads_strides_dilations[4];
+    filter.dilations[1] = test_case.pads_strides_dilations[5];
+
```
Insert:
```
     // A few basic dimension checks that we expect to be caught.
     bool valid_test_case =
                 wei_h > 2 * filter.pads[0] &&
                 wei_w > 2 * filter.pads[1] &&
                 input_h >= (2 * filter.pads[0] + wei_h) &&
                 input_w >= (2 * filter.pads[1] + wei_w)
     ;
```

kokolchin replied:
Implemented in 93a31d24d94b: added the explicit valid_test_case dimension checks block and switched the flow to use this gate directly.

randyspauldingamd commented on projects/miopen/test/gtest/cbna_inference.cpp:
```diff
@@ -0,0 +1,357 @@
+    int input_c, input_h, input_w, wei_c, wei_k, wei_h, wei_w;
+    std::tie(wei_k, wei_c, wei_h, wei_w)             = miopen::tien<4>(weights.desc.GetLengths());
+    std::tie(std::ignore, input_c, input_h, input_w) = miopen::tien<4>(input.desc.GetLengths());
+
```
Insert:
```
    EXPECT_EQ(input_c, wei_c) << "Aborting due to incorrect test config";
    if(input_c != wei_c)
    {
         return;
    }
```

kokolchin replied:
Implemented in commit e826bc4df48b: added EXPECT_EQ(input_c, wei_c) + early return.

randyspauldingamd commented on projects/miopen/test/gtest/cbna_inference.cpp:
```diff
@@ -0,0 +1,357 @@
+            ss << "CBNA Inference plan not supported for: " << test_case;
+            GTEST_SKIP() << ss.str();
+        }
+        else if(input.desc.GetLengths().at(1) == weights.desc.GetLengths().at(1) &&
```
After the changes in other comments, this `else` chain can be completely removed. The intent is that all failures and skips have been caught, so we can just call `VerifyAndValidate`, then exit.
I've tested on `gfx90a` and `gfx942`. Could you test on `navi` to verify this?

kokolchin replied:
The else if chain has been completely removed. Unfortunately, I don't have access to the Navi machine.

randyspauldingamd commented on projects/miopen/test/gtest/cbna_inference.cpp:
```diff
@@ -0,0 +1,357 @@
+        }
+        else
+        {
+            GTEST_SKIP() << "Test case dimensions do not meet requirements";
```
(should be) Unneeded. Remove.

kokolchin replied:
Removed.

randyspauldingamd reviewed (changes requested):
The port looks good--just the one request to revert some variable names.

I also requested some changes to `CBNA` to make it more of an integration test. I think any other refactoring or improvements should wait for future PRs. Hopefully my comments are complete and understandable. Feel free to reach out if I seem to have missed something, or with any other questions!

Finally, we do need to add name generators for all of these ctest conversions during the first PR. Please do so. Could you also check your other PRs, as well as remind the rest of the team? Thanks!

geomin12 reviewed (commented):
just allowed workflows to run for TheRock CI, awaiting results

please ping next time if you are waiting for approval for workflow run

kokolchin commented:
> The port looks good--just the one request to revert some variable names.
Done. 
> I also requested some changes to `CBNA` to make it more of an integration test. 
I think I did it. 
> Hopefully my comments are complete and understandable. 
Yes. Complete and understandable.

> Finally, we do need to add name generators for all of these ctest conversions during the first PR. Please do so. 
I added name generators, commit 374052f. 

> Could you also check your other PRs, as well as remind the rest of the team? 
Sure. My colleagues and I use name generators.

bethune-bryant commented:
Removing the `shared: origami` label — this PR only changes files under `projects/miopen/test/`, so origami isn't affected and we're unsure why this label was applied. Looking at the full label list, several other project labels on this PR look like they may also have been applied in error (the current `.github/labeler.yml` only applies project labels based on changed paths, which here would be just `project: miopen`). Worth a quick audit by the MIOpen team.

kokolchin commented:
> Finally, we do need to add name generators for all of these ctest conversions during the first PR. Please do so.

Done.

kokolchin commented:
> Removing the `shared: origami` label — this PR only changes files under `projects/miopen/test/`, so origami isn't affected and we're unsure why this label was applied. Looking at the full label list, several other project labels on this PR look like they may also have been applied in error (the current `.github/labeler.yml` only applies project labels based on changed paths, which here would be just `project: miopen`). Worth a quick audit by the MIOpen team.

Irrelevant labels removed.

Commit:
[MIOpen] Convert cba inference ctest to GTest.

Refresh PR scope on top of develop and keep only CBA/CBNA migration files.

Commit:
[MIOpen] Replace legacy cba_infer gtest with migrated implementation.

Keep the original cba_infer.cpp filename and remove cba_inference.cpp to avoid duplicate gtest symbols without changing test/gtest/CMakeLists.txt.

kokolchin commented:
Now I get a CI failure that appears to be infrastructure-side on the Windows runner, not test-content-related.
Failing step is workflow cleanup: build_tools/github_actions/cleanup_processes.ps1
with: PSSecurityException: running scripts is disabled on this system (UnauthorizedAccess), exit code 1.

This causes therock-ci-windows / TheRock CI Summary to fail.
Could CI maintainers please check runner execution policy or update the Windows shell invocation (e.g., process-scope bypass) for this cleanup step?
I can rerun once that is adjusted.
